docs: add task modifier explainer (#158)
mainmatter/sheepdog: apps/docs/src/content/docs/explainers/task-modifiers.mdx

		},
		{ max: 3 },
	);
</script>

<input
	type="file"
	multiple
	on:change={(e) => {
		files = [...e.target.files];
	}}
/>
<button
	on:click={() => {
		for (let file of files) {
			uploadFile.perform(file);
		}
	}}>start upload</button
>
```

You can play around with it in our Timeline and you may notice a new parameter available for you:
initially set at 1 you can update the `max` parameter to allow for more task to run at the same
time.

<Timeline selected_task_type="enqueue" client:load />

## Drop

The `drop` task will simply ignore (and thus drop) every `perform` that is executed when the `max`
amount of concurrent tasks is reached. In our example with a max concurrency of 1 for two seconds
after the first perform no other task will be accepted.

You can think of this kind of tasks as a sort of throttle and since the function will be just tossed
away you should use this modifier only when you don't care about the "execution loss". A good
example could be sending your mouse position to a websocket to be broadcasted to all the other
users. You don't want to update it too often to avoid clogging your server and you don't care about
the middle positions being lost.

```svelte

> <script lamng="ts">

Reviewer: ```suggestion
<script lang="ts">
```

	import { task, timeout } from '@sheepdogs/svelte';
	import ws from './websocket';

	const updatePosition = task.drop(async (coords: { x: number; y: number }) => {
		await timeout(200);
		ws.send('coords', coords);
	});
</script>

<svelte:window
	on:mousemove={(e) => {
		updatePosition.perform({
			x: e.clientX,
			y: e.clientY,
		});
	}}
/>
```

I know you are waiting for it so here's your `drop` Timeline playground:

<Timeline selected_task_type="drop" client:load />

## Restart

> The `restart` task will cancel the oldest task instance once the `max` is reached. With `max` of one
> this basically means that every time you perform again you will cancel the last task.

Reviewer: ```suggestion
The `restart` task will cancel the oldest task instance once the `max` is reached. With `max` of 1,
this basically means that every time you perform again you will cancel the last task.
```

> You can think of this kind of tasks as a the good 'ol debounce: if you await for, let's say, 200ms
> before doing any action in a `restart` task you can call `perform` repeatedly and the actual task
> will be executed only once the `perform` as not been called for 200ms. This is very useful to
> perform fetch requests while the user is typing without inundating your server of requests. Once the
> user stop typing the last request will not be canceled and the request will go through.

Reviewer: ```suggestion
You can think of this kind of tasks as a the good ol' debounce: if you await, let's say, 200ms
before doing any action in a `restart` task you can call `perform` repeatedly and the actual task
will be executed only once the `perform` as not been called for 200ms. This is very useful to
perform fetch requests while the user is typing without inundating your server with requests. Once the
user stops typing the last request will go through without being canceled.
```

```svelte

> <script lamng="ts">

Reviewer: ```suggestion
<script lang="ts">
```

	import { task, timeout } from '@sheepdogs/svelte';

	const search = task.restart(async (query: string) => {
		await timeout(200);
		const res = await fetch(`/api/search?q=${query}`);
	});
</script>

<input
	on:input={(e) => {
		search.perform(e.target.value);
	}}
/>
```

And as usual here's your `restart` Timeline playground:

<Timeline selected_task_type="restart" client:load />

## Keep Latest

> Finally `keepLatest` (we kept this latest _ba-dun-tss_) it's a mix of `drop` with a sprinkle of
> `enqueue`: once the max concurrency is reached it will drop every new `perform` but it will also
> keep the very last in queue to execute it when the max concurrency drop below the max.

Reviewer: ```suggestion
Finally, we arrive at `keepLatest` (we kept this latest _ba-dun-tss_). It's a mix of `drop` with a sprinkle of
`enqueue`: once the max concurrency is reached it will drop every new `perform` but it will also
keep the very last in the queue to execute when the max concurrency drop below the `max`.
```

"we kept this latest _ba-dun-tss_" is a nice pun but doesn't really make sense

Author: > "we kept this latest ba-dun-tss" is a nice pun but doesn't really make sense

Ok ok i will remove it 😦

description: How task modifiers work
---

import Timeline from '../TimelineExplainer.svelte';

In our [Getting started](/getting-started/usage#task-modifiers) you've seen that you can specify a
modifier on the task, either with the dot notation or the options notation. The reason you have this

> ability is so that you can specify what should happen when multiple, concurrent, task instance are

Reviewer 2: ```suggestion
ability is so that you can specify what should happen when multiple, concurrent task instances are
```

running.

> In this explainer we will guide you through how each of our modifiers work with examples of how they

Reviewer 2: ```suggestion
In this explainer, we will guide you through how each of our modifiers works with examples of how they
```

behave and hopefully make it clearer to understand when to use one or the other.

## Default

By default, every task simply runs as soon as you perform it. This is basically like a function call
with the added benefit of having derived state and cancellability. You might've seen in our home
page this simple demonstration; but if you didn't, allow me to introduce to you the friend that will
guide us through discovering the task modifiers: the Timeline demo!

Take your time to familiarize yourself with it: you can perform a new task by clicking on the

> `perform` button and reset the demo with the `clear timeline` button, the moment you perform a task

Reviewer 2: ```suggestion
`perform` button and reset the demo with the `clear timeline` button. The moment you perform a task,
```

the timeline will start and it will show you the state of the task. Each task will run for two
seconds and you can either click on it individually to cancel it or click on the `cancelAll` button
to nuke them all!

> <Timeline show_max={false} client:load />

Reviewer 2: I am not sure if the timeline behavior after all the tasks are finished or cancel is expected? It feels a bit weird to me.

https://github.com/user-attachments/assets/c960088e-5da6-4a39-a5ad-3f7eee7a36a7

Author: Yeah for the moment is expected.

## Enqueue

The `enqueue` task, as the name suggests, creates a queue of all the task instances after the max
concurrency has been reached. This means that up until the moment max concurrency is achieved every
task instance will be executed immediately. But if you try to perform a new task instance and the
number of running instances is at the `max` level, that new task instance will be stored in a queue
and executed as soon as the number of concurrent instances drops below the `max` again.

This is the right kind of task to use when you want to be sure that every task instance is executed
at some point (unless the user leaves the page) but you also don't want to run them concurrently. A
good use case for this could be uploading a list of files to the server. You want to upload all of
your files but you don't want to send them all together because that would be too resource intensive
and also makes it harder to handle in case of an error. You can use `enqueue` to allow only `max`
file uploads at a time.

```svelte
<script lang="ts">
	import { task, timeout } from '@sheepdogs/svelte';

> 	import upload from './file-upload';

Reviewer: Having this might be confusing, I'd probably opt for pseudo code instead

Reviewer: they do it this way in the svelte docs, let's just leave it

	let files = [];

	const uploadFile = task.enqueue(
		async (file: File) => {
			await upload(file);
		},
		{ max: 3 },
	);
</script>

<input
	type="file"
	multiple
	on:change={(e) => {
		files = [...e.target.files];
	}}
/>
<button
	on:click={() => {
		for (let file of files) {
			uploadFile.perform(file);
		}
	}}>start upload</button
>
```

You can play around with it in our Timeline and you may notice a new parameter available for you:
initially set at 1, you can update the `max` parameter to allow for more task instances to run at
the same time.

<Timeline selected_task_type="enqueue" client:load />

## Drop

The `drop` task will simply ignore (and thus drop) every `perform` that is executed when the `max`

> amount of concurrent task instances is reached. In our example with a max concurrency of 1 for two
> seconds after the first perform no other task instance will be executed.

Reviewer: ```suggestion
amount of concurrent task instances is exceeded. With a max concurrency of 1, our example will drop any other tasks that are performed while our initial task is running.
```

You can think of this kind of task as a sort of throttle; and since the function will be just tossed
away you should use this modifier only when you don't care about the "execution loss". A good
example could be sending your mouse position to a websocket to be broadcasted to all the other
users. You don't want to update it too often to avoid clogging your server, you don't care about the

> middle positions being lost, and the final mouse position is the most important.

Reviewer: In this specific example, I would have thought `keepLatest` would be the best option.

```svelte
<script lang="ts">
	import { task, timeout } from '@sheepdogs/svelte';
	import ws from './websocket';

	const updatePosition = task.drop(async (coords: { x: number; y: number }) => {
		await timeout(200);
		ws.send('coords', coords);
	});
</script>

<svelte:window
	on:mousemove={(e) => {
		updatePosition.perform({
			x: e.clientX,
			y: e.clientY,
		});
	}}
/>
```

I know you are waiting for it so here's your `drop` Timeline playground:

<Timeline selected_task_type="drop" client:load />

## Restart

> The `restart` task will cancel the oldest task instance once the `max` is reached. With `max` of 1,

Reviewer: ```suggestion
The `restart` task will cancel the oldest task instance once the `max` is exceeded. With `max` of 1,
```

> this basically means that every time you perform again you will cancel the last task.

Reviewer: ```suggestion
this basically means that every time you perform again you will cancel the last task instance.
```

> You can think of this kind of tasks as a the good ol' debounce: if you await for, let's say, 200ms

Reviewer: ```suggestion
You can think of this kind of task as a the good ol' debounce: if you await for, let's say, 200ms
```

before doing any action in a `restart` task you can call `perform` repeatedly and the actual task
will be executed only once the `perform` as not been called for 200ms. This is very useful to
perform fetch requests while the user is typing without inundating your server with requests. Once
the user stops typing the last request will go through without being canceled.

```svelte
<script lang="ts">
	import { task, timeout } from '@sheepdogs/svelte';

	const search = task.restart(async (query: string) => {
		await timeout(200);
		const res = await fetch(`/api/search?q=${query}`);
	});
</script>

<input
	on:input={(e) => {
		search.perform(e.target.value);
	}}
/>
```

> And as usual here's your `restart` Timeline playground:

Reviewer: ```suggestion
And as per usual, here's your `restart` Timeline playground:
```

<Timeline selected_task_type="restart" client:load />

## Keep Latest

Finally, we arrive at `keepLatest`. It's a mix of `drop` with a sprinkle of `enqueue`: once the max

> concurrency is reached it will drop every new `perform` but it will also keep the very last in the

Reviewer: ```suggestion
concurrency is exceeded it will drop every new `perform` but it will also keep the very last in the
```

notification that allow you to reload the document data with the newest additions. If there are two
or more saves while you are fetching the data you:

1. don't want to restart the current fetch because you want to show it to the user immediately
2. don't want to drop the new task because you actually want the latest data
3. don't case about the in-between data because the last one is the actual content of the file

That's where `keepLatest` will come in handy!

```svelte
<script lang="ts">
	import Document from '$lib/Document.svelte';
	import { ws } from '$lib/websocket';
	import { task } from '@sheepdogs/svelte';
	import { onMount } from 'svelte';

	let document;

	const loadDocument = task.keepLatest(async () => {
		const res = await fetch(`/api/get-document`);
		document = await res.json();
	});

	onMount(() => {
		loadDocument.perform();
		return ws.on('new-save', () => {
			loadDocument.perform();
		});
	});
</script>

<Document {document} />
```

<Timeline selected_task_type="keepLatest" client:load />

## Conclusion

Task modifiers are very powerful and one of the core features of `@sheepdog/svelte`, this explainer

> aim to give you all the tools to use them at their best...and some playful time with the timeline

Reviewer: ```suggestion
aims to give you all the tools to use them at their best...and some playful time with the timeline
```
